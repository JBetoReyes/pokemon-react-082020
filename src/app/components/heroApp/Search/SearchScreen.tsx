import React, { useMemo, useState } from 'react';
import useForm from '@hooks/useForm';
import { AppSubmitEvent, AppChangeEvent } from '@typings/htmlEvents';
import { useLocation, RouteComponentProps } from 'react-router-dom';
import { parse } from 'query-string';
import getHeroeByName from '../services/getHeroeByName';
import HeroCard from '../heroes/HeroCard';

export type Props = RouteComponentProps;

const SearchScreen = ({ history }: Props): JSX.Element => {
  const [searchTriggered, setSearchState] = useState(false);
  const { search } = useLocation();
  const { q = '' } = parse(search);
  const { formData, handleChange } = useForm({ query: `${q}` });
  const { query } = formData;
  const heroesFiltered = getHeroeByName(q as string);
  const handleSearchChange = (e: AppChangeEvent) => {
    setSearchState(false);
    handleChange(e);
  };
  const handleSubmit = (e: AppSubmitEvent) => {
    setSearchState(true);
    e.preventDefault();
    history.push(`?q=${query}`);
  };
  return (
    <div>
      <div className="container-fuid">
        <h1 className="pl-5">Search SearchScreen</h1>
        <hr />
        <div className="row pl-5">
          <div className="col-4">
            <h4>Search Form</h4>
            <hr />
            <form onSubmit={handleSubmit}>
              <input
                type="text"
                placeholder="Find your hero"
                className="form-control"
                name="query"
                autoComplete="off"
                value={query}
                onChange={handleSearchChange}
              />
              <button
                type="submit"
                className="btn btn-outline-primary btn-block mt-3"
              >
                Search...
              </button>
            </form>
          </div>
          <div className="col-7">
            <h4>Results</h4>
            <hr />
            {query === '' && heroesFiltered.length === 0 && (
              <div className="alert alert-info ml-3 mr-3">Search a heroe</div>
            )}
            {query !== '' && heroesFiltered.length === 0 && searchTriggered && (
              <div className="alert alert-danger ml-3 mr-3">
                {`Theres is no a hero with ${query}`}
              </div>
            )}
            {heroesFiltered.map((heroe) => (
              <div className="m-3">
                <HeroCard key={heroe.id} {...heroe} />
              </div>
            ))}
          </div>
        </div>
      </div>
    </div>
  );
};

export default SearchScreen;
