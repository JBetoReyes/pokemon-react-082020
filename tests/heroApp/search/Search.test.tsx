/* eslint-disable @typescript-eslint/no-explicit-any */
import SearchScreen from '@components/heroApp/Search/SearchScreen';
import { MemoryRouter, Route } from 'react-router-dom';
import React from 'react';
import { mount } from 'enzyme';
import { act } from 'react-dom/test-utils';

describe('Search tests', () => {
  let mockHistory;
  beforeEach(() => {
    mockHistory = {
      push: jest.fn(),
    };
  });
  afterAll(() => {
    jest.resetAllMocks();
  });
  test('Should match the snapshot', () => {
    const wrapper = mount(
      <MemoryRouter initialEntries={['/search']}>
        <Route
          path="/search"
          component={() => <SearchScreen history={mockHistory} />}
        />
      </MemoryRouter>
    );
    expect(wrapper.find('.container-fluid')).toMatchSnapshot();
  });
  test('Should received the query params', () => {
    const wrapper = mount(
      <MemoryRouter initialEntries={['/search?q=batman']}>
        <Route
          path="/search"
          component={() => <SearchScreen history={mockHistory} />}
        />
      </MemoryRouter>
    );
    expect(wrapper.find('SearchScreen')).toMatchSnapshot();
    expect(wrapper.find('input').prop('value')).toBe('batman');
  });
  test('Should render the warning component', () => {
    const wrapper = mount(
      <MemoryRouter initialEntries={['/search?q=noresult']}>
        <Route
          path="/search"
          component={() => <SearchScreen history={mockHistory} />}
        />
      </MemoryRouter>
    );
    act(() => {
      wrapper.find('form').prop('onSubmit')({ preventDefault() {} } as any);
    });
    wrapper.mount();
    expect(wrapper.find('.alert').exists()).toBe(true);
  });
  test('Should call the push function', () => {
    const wrapper = mount(
      <MemoryRouter initialEntries={['/search?q=batman1234']}>
        <Route
          path="/Search"
          component={() => <SearchScreen history={mockHistory} />}
        />
      </MemoryRouter>
    );
    act(() => {
      wrapper.find('input').simulate('change', {
        target: {
          name: 'query',
          value: 'batman',
        },
      });
    });
    wrapper.update();
    act(() => {
      wrapper.find('form').prop('onSubmit')({ preventDefault() {} } as any);
    });
    expect(mockHistory.push).toBeCalledWith('?q=batman');
  });
});
