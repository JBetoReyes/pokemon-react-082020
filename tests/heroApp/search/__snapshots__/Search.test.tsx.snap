// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Search tests Should match the snapshot 1`] = `
<div
  className="container-fluid"
>
  <h1
    className="pl-5"
  >
    Search SearchScreen
  </h1>
  <hr />
  <div
    className="row pl-5"
  >
    <div
      className="col-4"
    >
      <h4>
        Search Form
      </h4>
      <hr />
      <form
        onSubmit={[Function]}
      >
        <input
          autoComplete="off"
          className="form-control"
          name="query"
          onChange={[Function]}
          placeholder="Find your hero"
          type="text"
          value=""
        />
        <button
          className="btn btn-outline-primary btn-block mt-3"
          type="submit"
        >
          Search...
        </button>
      </form>
    </div>
    <div
      className="col-7"
    >
      <h4>
        Results
      </h4>
      <hr />
      <div
        className="alert alert-info ml-3 mr-3"
      >
        Search a heroe
      </div>
    </div>
  </div>
</div>
`;

exports[`Search tests Should received the query params 1`] = `
<SearchScreen
  history={
    Object {
      "push": [MockFunction],
    }
  }
>
  <div>
    <div
      className="container-fluid"
    >
      <h1
        className="pl-5"
      >
        Search SearchScreen
      </h1>
      <hr />
      <div
        className="row pl-5"
      >
        <div
          className="col-4"
        >
          <h4>
            Search Form
          </h4>
          <hr />
          <form
            onSubmit={[Function]}
          >
            <input
              autoComplete="off"
              className="form-control"
              name="query"
              onChange={[Function]}
              placeholder="Find your hero"
              type="text"
              value="batman"
            />
            <button
              className="btn btn-outline-primary btn-block mt-3"
              type="submit"
            >
              Search...
            </button>
          </form>
        </div>
        <div
          className="col-7"
        >
          <h4>
            Results
          </h4>
          <hr />
          <div
            className="m-3"
            key="dc-batman"
          >
            <HeroCard
              alter_ego="Bruce Wayne"
              characters="Bruce Wayne"
              first_appearance="Detective Comics #27"
              id="dc-batman"
              publisher="DC Comics"
              superhero="Batman"
            >
              <div
                className="card ms-3"
                style={
                  Object {
                    "maxWidth": 540,
                  }
                }
              >
                <div
                  className="row no-gutters"
                >
                  <div
                    className="col-md-4"
                  >
                    <img
                      alt="Batman"
                      className="card-img"
                      src="./heroes/dc-batman.jpg"
                    />
                  </div>
                  <div
                    className="col-md-8"
                  >
                    <div
                      className="card-body"
                    >
                      <h5
                        className="card-title"
                      >
                        Batman
                      </h5>
                      <p
                        className="card-text"
                      >
                        Bruce Wayne
                      </p>
                      <p
                        className="card-text"
                      >
                        <small
                          className="text-muted"
                        >
                          Detective Comics #27
                        </small>
                      </p>
                      <Link
                        to="./hero/dc-batman"
                      >
                        <LinkAnchor
                          href="/hero/dc-batman"
                          navigate={[Function]}
                        >
                          <a
                            href="/hero/dc-batman"
                            onClick={[Function]}
                          >
                            More...
                          </a>
                        </LinkAnchor>
                      </Link>
                    </div>
                  </div>
                </div>
              </div>
            </HeroCard>
          </div>
        </div>
      </div>
    </div>
  </div>
</SearchScreen>
`;

exports[`Search tests Should render the warning component 1`] = `
<MemoryRouter
  initialEntries={
    Array [
      "/search?q=noresult",
    ]
  }
>
  <Router
    history={
      Object {
        "action": "POP",
        "block": [Function],
        "canGo": [Function],
        "createHref": [Function],
        "entries": Array [
          Object {
            "hash": "",
            "key": "5szmuy",
            "pathname": "/search",
            "search": "?q=noresult",
            "state": undefined,
          },
        ],
        "go": [Function],
        "goBack": [Function],
        "goForward": [Function],
        "index": 0,
        "length": 1,
        "listen": [Function],
        "location": Object {
          "hash": "",
          "key": "5szmuy",
          "pathname": "/search",
          "search": "?q=noresult",
          "state": undefined,
        },
        "push": [Function],
        "replace": [Function],
      }
    }
  >
    <Route
      component={[Function]}
      path="/search"
    >
      <component
        history={
          Object {
            "action": "POP",
            "block": [Function],
            "canGo": [Function],
            "createHref": [Function],
            "entries": Array [
              Object {
                "hash": "",
                "key": "5szmuy",
                "pathname": "/search",
                "search": "?q=noresult",
                "state": undefined,
              },
            ],
            "go": [Function],
            "goBack": [Function],
            "goForward": [Function],
            "index": 0,
            "length": 1,
            "listen": [Function],
            "location": Object {
              "hash": "",
              "key": "5szmuy",
              "pathname": "/search",
              "search": "?q=noresult",
              "state": undefined,
            },
            "push": [Function],
            "replace": [Function],
          }
        }
        location={
          Object {
            "hash": "",
            "key": "5szmuy",
            "pathname": "/search",
            "search": "?q=noresult",
            "state": undefined,
          }
        }
        match={
          Object {
            "isExact": true,
            "params": Object {},
            "path": "/search",
            "url": "/search",
          }
        }
      >
        <SearchScreen
          history={
            Object {
              "push": [MockFunction] {
                "calls": Array [
                  Array [
                    "?q=noresult",
                  ],
                ],
                "results": Array [
                  Object {
                    "type": "return",
                    "value": undefined,
                  },
                ],
              },
            }
          }
        >
          <div>
            <div
              className="container-fluid"
            >
              <h1
                className="pl-5"
              >
                Search SearchScreen
              </h1>
              <hr />
              <div
                className="row pl-5"
              >
                <div
                  className="col-4"
                >
                  <h4>
                    Search Form
                  </h4>
                  <hr />
                  <form
                    onSubmit={[Function]}
                  >
                    <input
                      autoComplete="off"
                      className="form-control"
                      name="query"
                      onChange={[Function]}
                      placeholder="Find your hero"
                      type="text"
                      value="noresult"
                    />
                    <button
                      className="btn btn-outline-primary btn-block mt-3"
                      type="submit"
                    >
                      Search...
                    </button>
                  </form>
                </div>
                <div
                  className="col-7"
                >
                  <h4>
                    Results
                  </h4>
                  <hr />
                  <div
                    className="alert alert-danger ml-3 mr-3"
                  >
                    Theres is no a hero with noresult
                  </div>
                </div>
              </div>
            </div>
          </div>
        </SearchScreen>
      </component>
    </Route>
  </Router>
</MemoryRouter>
`;
